pkg/registry/rbac/clusterrole/policybased/storage_test.go

```diff
@@ -0,0 +1,253 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policybased
+
+import (
+	"context"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	_ "k8s.io/kubernetes/pkg/apis/rbac/install"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+)
+
+func TestEscalation(t *testing.T) {
+	createContext := request.WithRequestInfo(request.WithNamespace(context.TODO(), ""), &request.RequestInfo{
+		IsResourceRequest: true,
+		Verb:              "create",
+		APIGroup:          "rbac.authorization.k8s.io",
+		APIVersion:        "v1",
+		Namespace:         "",
+		Resource:          "clusterroles",
+		Name:              "",
+	})
+	updateContext := request.WithRequestInfo(request.WithNamespace(context.TODO(), ""), &request.RequestInfo{
+		IsResourceRequest: true,
+		Verb:              "update",
+		APIGroup:          "rbac.authorization.k8s.io",
+		APIVersion:        "v1",
+		Namespace:         "",
+		Resource:          "clusterroles",
+		Name:              "myrole",
+	})
+
+	superuser := &user.DefaultInfo{Name: "superuser", Groups: []string{"system:masters"}}
+	superuserCreateCtx := request.WithUser(createContext, superuser)
+	superuserUpdateCtx := request.WithUser(updateContext, superuser)
+
+	bob := &user.DefaultInfo{Name: "bob"}
+	bobCreateCtx := request.WithUser(createContext, bob)
+	bobUpdateCtx := request.WithUser(updateContext, bob)
+
+	alice := &user.DefaultInfo{Name: "alice"}
+	aliceCreateCtx := request.WithUser(createContext, alice)
+	aliceUpdateCtx := request.WithUser(updateContext, alice)
+
+	var (
+		authzCalled   = 0
+		authzDecision = authorizer.DecisionNoOpinion
+		authzReason   string
+		authzErr      error
+	)
+	fakeStorage := &fakeStorage{}
+	fakeAuthorizer := authorizer.AuthorizerFunc(func(attr authorizer.Attributes) (authorizer.Decision, string, error) {
+		authzCalled++
+		return authzDecision, authzReason, authzErr
+	})
+	fakeRuleResolver, _ := validation.NewTestRuleResolver(
+		nil,
+		nil,
+		[]*rbac.ClusterRole{{ObjectMeta: metav1.ObjectMeta{Name: "alice-role"}, Rules: []rbac.PolicyRule{{APIGroups: []string{"*"}, Resources: []string{"*"}, Verbs: []string{"*"}}}}},
+		[]*rbac.ClusterRoleBinding{{RoleRef: rbac.RoleRef{Name: "alice-role", APIGroup: "rbac.authorization.k8s.io", Kind: "ClusterRole"}, Subjects: []rbac.Subject{{Name: "alice", Kind: "User", APIGroup: "rbac.authorization.k8s.io"}}}},
+	)
+
+	role := &rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: "myrole", Namespace: "", Labels: map[string]string{"invalid name": ""}},
```

Comment:
What's this label for?

Reply:
leftover from an early attempt at using generic storage and preventing writes by failing validation. abandoned that in favor of a fake storage object. removed.